src/sinks/aws_s3/sink.rs

```diff
@@ -0,0 +1,521 @@
+        batch_timeout: Duration,
+        options: S3RequestOptions,
+    ) -> Self {
+        Self {
+            acker: Some(cx.acker()),
+            service: Some(service),
+            partitioner: Some(partitioner),
+            batch_size_bytes,
+            batch_size_events,
+            batch_timeout,
+            options,
+        }
+    }
+}
+
+#[async_trait]
+impl<S> StreamSink for S3Sink<S>
+where
+    S: Service<S3Request> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Response: Response + Send + 'static,
+    S::Error: Debug + Into<crate::Error> + Send,
+{
+    async fn run(&mut self, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        // All sinks do the same fundamental job: take in events, and ship them
+        // out. Empirical testing shows that our number one priority for high
+        // throughput needs to be servicing I/O as soon as we possibly can.  In
+        // order to do that, we'll spin up a separate task that deals
+        // exclusively with I/O, while we deal with everything else here:
+        // batching, ordering, and so on.
+        let (io_tx, io_rx) = channel(64);
+        let service = self
+            .service
+            .take()
+            .expect("same sink should not be run twice");
+        let acker = self
+            .acker
+            .take()
+            .expect("same sink should not be run twice");
+        let partitioner = self
+            .partitioner
+            .take()
+            .expect("same sink should not be run twice");
+
+        let io = run_io(io_rx, service, acker).in_current_span();
+        let _ = tokio::spawn(async move {
+            io.await;
+        });
+
+        let batcher = Batcher::new(
+            input,
+            partitioner,
+            self.batch_timeout,
+            self.batch_size_bytes,
+            self.batch_size_events,
+        );
+        pin!(batcher);
+
+        while let Some((key, batch)) = batcher.next().await {
+            match key {
+                Some(key) => {
+                    // We could push this down to the I/O task if we wanted to.
+                    let request = build_request(key, batch, &self.options);
+                    if io_tx.send(request).await.is_err() {
+                        // TODO: change this to "error! + return Err" after
+                        // initial testing/debugging
+                        trace!(
+                            "Sink I/O channel should not be closed before sink itself is closed."
+                        );
+                        panic!("boom 1")
+                    }
+                }
+                // Partitioning failed for one or more events.
+                //
+                // TODO: this might be where we would insert something like the
+                // proposed error handling/dead letter queue stuff; events that
+                // _we_ can't handle, but some other component may be able to
+                // salvage
```

Comment:
When would this fail in practice? Currently we drop any events that cannot be templated.

Reply:
It will fail if the template references a field that doesn't exist in the log, seems to me. The partitioner currently kicks out a `TemplateRenderingFailed` so we don't drop this silently (even if the message ends up being non-local to this code).

Reply:
Exactly.